### PR TITLE
Add S3Path constructor that takes in S3Path and version

### PR DESCRIPTION
**Pull request by hannahilea**
Closes https://github.com/JuliaCloud/AWSS3.jl/issues/294.

In this PR, fails if input path already has a version; could alternatively implement silently overwriting an existing version.

### CODE REVIEW

**Comment by ararslan** [comment]
> could alternatively implement silently overwriting an existing version

FWIW this is the behavior I would expect as a user. The same is true of `config`. It'd be nice more generally if `S3Path(::S3Path; settings...)` gave you back an `S3Path` with the provided `settings`, overwriting or populating as the case may be.

**Comment by hannahilea** [comment]
> FWIW this is the behavior I would expect as a user.

Iiiiiiinteresting. 

>  It'd be nice more generally if...

This was actually the path I originally started down, before deciding that failure was the desired end state. Hrm.

**Review by mattBrzezinski — APPROVED** [review]

File: src/s3path.jl [hunk 1]
```diff
@@ -46,6 +46,7 @@ end
 
 """
     S3Path()
+    S3Path(path::S3Path; version::$(AbstractS3Version))
     S3Path(str; version::$(AbstractS3Version)=nothing, config::$(AbstractS3PathConfig)=nothing)
```
Comment (omus) [inline src/s3path.jl:50]:
```suggestion
    S3Path(str; version::$(AbstractS3Version)=nothing, config::$(AbstractS3PathConfig)=nothing)
    S3Path(path::S3Path; version::$(AbstractS3Version))
```

File: src/s3path.jl [hunk 2]
```diff
@@ -107,6 +108,17 @@ function S3Path(
     )
 end
 
+function S3Path(path::S3Path; version::AbstractS3Version)
+    if !isnothing(path.version)
+        throw(
+            ArgumentError(
+                "Can't construct versioned path, input already specifies version: $path"
```
Comment (omus) [inline src/s3path.jl:115]: Why not allow the keyword version to overwrite the version of the `S3Path` silently? Should probably document the rational for this as part of the docstring

**Comment by omus** [comment]
bors try

**Comment by hannahilea** [comment]
Okay, I updated the approach here to be in line with @ararslan's expectations. Rereview required!

File: src/s3path.jl [hunk 3]
```diff
@@ -46,10 +46,12 @@ end
 
 """
     S3Path()
-    S3Path(str; version::$(AbstractS3Version)=nothing, config::$(AbstractS3PathConfig)=nothing)
+    S3Path(str::AbstractString; version::$(AbstractS3Version)=nothing, config::$(AbstractS3PathConfig)=nothing)
+    S3Path(path::S3Path; isdirectory=missing, version=missing, config=missing)
```
Comment (ararslan) [inline src/s3path.jl:50]: Why `missing` in this case but `nothing` in the others? Note that a simpler way to go about this is to reference the fields of the positional argument as the default values for the keyword arguments, e.g.:
```julia
function S3Path(path::S3Path; isdirectory=path.isdirectory, version=path.version,
                config=path.config)
    return S3Path(path.bucket, path.key; isdirectory, version, config)
end
```

Comment (hannahilea) [reply src/s3path.jl:50]: oh i love this

Comment (hannahilea) [reply src/s3path.jl]: i'm annoyed that i didn't come up with this myself, lol

**Comment by ararslan** [comment]
bors try

**Comment by ararslan** [comment]
bors try

File: CHANGELOG.md [hunk 4]
```diff
@@ -8,6 +8,8 @@
 
 - v0.11.0: The `s3_exists` and `isdir(::S3Path)` calls no longer encounter HTTP 403 (Access Denied) errors when attempting to list resources which requiring an `s3:prefix` to be specified ([#289]).
 - v0.11.1: The new keyword argument `returns` for `Base.write(fp::S3Path, ...)` determines the output returned from `write`, which can now be the raw `AWS.Response` (`returns=:response`) or the `S3Path` (`returns=:path`); this latter option returns an `S3Path` populated with the version ID of the written object (when versioning is enabled on the bucket) ([#293]).
+- v0.11.x: Add constructor for constructing `S3Path` from another `S3Path` ([#297]).
```
Comment (ararslan) [inline CHANGELOG.md:11]:
```suggestion
- v0.11.2: Add constructor for constructing `S3Path` from another `S3Path` ([#297]).
```

**Review by ararslan — APPROVED** [review]

File: test/s3path.jl [hunk 5]
```diff
@@ -636,6 +636,34 @@ function s3path_tests(base_config)
         @test_throws ArgumentError S3Path("s3://my_bucket/"; version="")
     end
 
+    @testset "construct s3path from s3path" begin
```
Comment (omus) [inline test/s3path.jl:639]:
```suggestion
    @testset "construct S3Path from S3Path" begin
```

File: test/s3path.jl [hunk 6]
```diff
@@ -636,6 +636,34 @@ function s3path_tests(base_config)
...
+        version2 = String('0':'5') * String('A':'Z')
+        @test S3Path(path_versioned; version=version2).version == version2 != version
+
+        # config kwarg overrides path.config
+        config2 = AWSConfig(; region="foo")
+        @test S3Path(path; config=config2).config == config2 != path.config
+        @test isnothing(S3Path(path; config=nothing).config)
+
+        # isdirectory kwarg overrides path.config
+        @test S3Path(path; isdirectory=!path.isdirectory) != path.isdirectory
```
Comment (omus) [inline test/s3path.jl:664]: Refactor of the tests to make them more consistent:
```suggestion
        version = String('A':'Z') * String('0':'5')
        p = S3Path(path; version)
        @test p != path
        @test p.bucket == path.bucket
        @test p.key == path.key
        @test p.config == path.config
        @test p.version == version != path.version

        # ...if version already exists, overwrite silently
        path_versioned = S3Path(path; version)
        alt_version = String('0':'5') * String('A':'Z')
        p = S3Path(path_versioned; version=alt_version)
        @test p.version == alt_version != path_versioned.version

        # config kwarg overrides path.config
        alt_config = AWSConfig(; region="foo")
        p = S3Path(path; config=alt_config)
        @test p.config == alt_config != path.config

        # isdirectory kwarg overrides path.config
        p = S3Path(path; isdirectory=!path.isdirectory)
        @test p.isdirectory != path.isdirectory
```

File: test/s3path.jl [hunk 7]
```diff
@@ -636,6 +636,34 @@ function s3path_tests(base_config)
...
+        path = S3Path("s3://my_bucket/prefix"; config)
+
+        # When no kwargs provided, return identity
+        @test S3Path(path) === path
+
+        # version kwarg overrides path.version
+        version = String('A':'Z') * String('0':'5')
+        path_versioned = S3Path(path; version)
+        @test path_versioned.version == version
+        @test path_versioned == S3Path("s3://my_bucket/prefix"; version)
```
Comment (omus) [inline test/s3path.jl:651]: Note that `config` isn't being used in the comparison here

File: CHANGELOG.md [hunk 8]
```diff
@@ -8,6 +8,8 @@
 
 - v0.11.0: The `s3_exists` and `isdir(::S3Path)` calls no longer encounter HTTP 403 (Access Denied) errors when attempting to list resources which requiring an `s3:prefix` to be specified ([#289]).
 - v0.11.1: The new keyword argument `returns` for `Base.write(fp::S3Path, ...)` determines the output returned from `write`, which can now be the raw `AWS.Response` (`returns=:response`) or the `S3Path` (`returns=:path`); this latter option returns an `S3Path` populated with the version ID of the written object (when versioning is enabled on the bucket) ([#293]).
+- v0.11.2: Add constructor for constructing `S3Path` from another `S3Path` ([#297]).
```
Comment (omus) [inline CHANGELOG.md:11]:
```suggestion
- v0.11.2: Add `S3Path` copy constructor for allowing updating `version`, `config`, and/or `isdirectory` ([#297]).
```

**Review by omus — APPROVED** [review]
Last minor changes

**Pull request merged**